docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,552 @@
+==============
+
+A :code:`let` binding creates a named variable that is bound to the given value
+and scoped to the subsequent expression. By contrast, a graph binding allows for
+explicitly constructing dataflow graphs in a Relay program by binding an expression
+(graph node) directly to a temporary variable, which is not scoped. Each reference
+to the variable corresponds to an edge in the dataflow graph. This has the
+semantics of substituting the expression wherever the variable appears, even though
+the graph node will only be evaluated once by the compiled program.
+
+These bindings allow for a style of programming that corresponds to that already
+employed by NNVM and other dataflow graph-based input formats. The fact that the variables
+are not scoped offers some flexibility in evaluation order compared to :code:`let`
+bindings, though this can also introduce some ambiguity in programs (the
+`developer introduction to the Relay IR`__ includes more detailed discussion
+of this nuance).
+
+__ `Introduction to Relay IR`_
+
+.. *Note: Graph bindings are not currently parsed by the text format.*
+
+In Relay's text format, a graph binding can be written as below (note the lack of a
+:code:`let` keyword and a semicolon):
+
+.. code-block:: python
+
+   %1 = %a + %b
+   %2 = %1 + %1
+   %2 * %2
+
+Unlike a let binding, a graph binding is not represented as an AST node in Relay, but rather as a meta-variable referencing its AST node value.
+For example, a program like the above could be constructed in Relay's
+Python front-end by setting *Python variables* equal to the corresponding Relay AST node and
+using the variables repeatedly, as below (a C++ program using the corresponding API bindings
+could accomplish the same thing):
+
+.. code-block:: python
+
+   sum1 = relay.add(a, b)
+   sum2 = relay.add(sum1, sum1)
+   relay.multiply(sum2, sum2)
+
+For development purposes and to enable certain optimizations, Relay includes passes to
+convert between dataflow graphs defined using graph bindings and programs with :code:`let`
+bindings in A-normal form, employed by many compiler optimizations from the functional
+programming community (see `"A-Normalization: Why and How" by
+Matt Might<http://matt.might.net/articles/a-normalization/>`__ for an introduction
+to the A-normal form).
+
+If-Then-Else
+============
+
+Relay has a simple if-then-else expression that allows programs to branch
+on a single value of type :code:`bool`, i.e., a zero-rank
+tensor of booleans (:code:`Tensor[(), bool]`).
+
+.. code-block:: python
+
+    if (equal(%t, %u)) {
+        %t
+    } else {
+        %u
+    }
+
+Since if-then-else branches are expressions, they may appear inline
+wherever any other expression may be expected, like invocations of
+the ternary operator in C-like languages. The if-then-else expression
+evaluates to the value of the "then" branch if the condition value
+evaluates to `True` and evaluates to the value of the "else" branch otherwise.
```

Comment:
I think there is some disagreement over whether the semantics are truthy-falsy or true-false, at least from some of the code I've read. @jroesch @MarisaKirisame.

Reply:
Yes, we should make a normative decision on that.

Reply:
I believe it is true-false. We should rephrase this sentence so False is the only value that triggers execution of the else branch.

Reply:
Okay, will do